### PR TITLE
Set server name only for the current broker

### PR DESCRIPTION
**Pull request by d1egoaz**
Fixes https://github.com/Shopify/sarama/issues/1700

### CODE REVIEW

File: broker.go [hunk 1]
```diff
@@ -1440,3 +1422,25 @@ func (b *Broker) registerCounter(name string) metrics.Counter {
 	b.registeredMetrics = append(b.registeredMetrics, nameForBroker)
 	return metrics.GetOrRegisterCounter(nameForBroker, b.conf.MetricRegistry)
 }
+
+func validServerNameTLS(addr string, conf *tls.Config) *tls.Config {
+	cfg := conf
```
Comment (dselyuzhitskiy) [inline broker.go:1427]: @d1egoaz i think here we don't need separate variable and could use conf from args just checked if it nil

**Comment by dnwe** [comment]
@d1egoaz rather than the special splitting code, can't we just use [net.SplitHostPort](https://golang.org/pkg/net/#SplitHostPort) ?

Also, I can't think of a good reason why anyone would have set a single ServerName in their `conf.Net.TLS.Config` to use for all brokers, so I wonder if we should always override whatever is there by forming it with our hostname.

So I wonder if we'd just do something like this:

```go
	var cfg *tls.Config
	if config.Net.TLS.Config == nil {
		cfg = &tls.Config{}
	} else {
		cfg = config.Net.TLS.Config.Clone()
	}

	if cfg.ServerName, _, err = net.SplitHostPort(broker); err != nil {
		return fmt.Errorf("failed to extract SNI got %w", err)
	}

	b.conn = tls.Client(b.conn, cfg)
```

**Comment by d1egoaz** [comment]
> @d1egoaz rather than the special splitting code, can't we just use net.SplitHostPort ?

for sure! thanks!

> Also, I can't think of a good reason why anyone would have set a single ServerName in their conf.Net.TLS.Config to use for all brokers, so I wonder if we should always override whatever is there by forming it with our hostname.

yeah, but I wonder if we would break any other client for doing this, looks like a breaking change.

**Comment by d1egoaz** [comment]
@dnwe @dselyuzhitskiy let me know what you guys think about the latest commits, and if it's approved I'll create a new release today.

I'm testing this branch in one of my apps, and no issues so far

File: broker.go [hunk 2]
```diff
@@ -1440,3 +1422,20 @@ func (b *Broker) registerCounter(name string) metrics.Counter {
...
+		cfg = &tls.Config{}
+	}
+	if cfg.ServerName != "" {
+		return cfg
+	}
+
+	c := cfg.Clone()
+	sn, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		Logger.Println(fmt.Errorf("failed to get ServerName from addr %w", err))
```
Comment (dselyuzhitskiy) [inline broker.go:1437]: Should we return with error here?

**Comment by dselyuzhitskiy** [comment]
Looks good, but suggestion from @dnwe is also interesting

File: broker.go (hunk 2 above)
Comment (d1egoaz) [reply broker.go:1437]: I should have mentioned this, as this was on purpose.

Currently, this code is called inside a go routine:
https://github.com/Shopify/sarama/blob/master/broker.go#L154

and currently the pattern inside this routine is to log the errors as there is no way to return an error to the caller.

**Comment by d1egoaz** [comment]
@dnwe let us know what do you think about the latest changes.
I'm not sure about overriding the user's configuration, looks like this change might be for v2

**Comment by dnwe** [comment]
@d1egoaz lets go ahead and merge and get another minor version bump out 👍🏻

**Pull request merged**